### ISSUE TITLE
explicitly specify stagedir

### ISSUE THREAD

**Issue by pohly** [clearlinux/swupd-client#273]
By default, swupd stages content under the statedir, which is not useful for a statedir that is not on the same partition as the target path, because then staged files need to be copied once more. See https://github.com/clearlinux/swupd-client/issues/249#issuecomment-329431119

Please add an explicit --stagedir which overrides the default.

**Comment by phmccarty** [comment]
@pohly Since swupd currently has two staging phases, can you clarify what --stagedir should refer to? The phases are (a) extracting fullfiles and applying deltas to populate the content of STATEDIR/staged, and (b) hardlinking (or copying) the appropriate STATEDIR/staged files to hidden files at the destination path under PATHPREFIX/.

Step (b) occurs in the `do_staging()` function, and step (a) manipulates the STATEDIR/staged directory, so the terminology is a bit confusing at the moment.

**Comment by phmccarty** [comment]
On second thought, I'm guessing you mean step (a) as I describe above, because there are usually multiple directories involved in step (b).

**Comment by pohly** [comment]
The goal is to implement step (b) using hardlinking instead of having to copy.

My understanding is that letting --stagedir override the STATEDIR/staged default will achive that if a directory is chosen on the target partition.

**Comment by icarus-sparry** [comment]
With lots of assumptions, like there is only one target partition.

This assumption is invalid in one of the things I am looking at, so it has to have a fallback to allow copying.
Would not manually creating a symlink for STATEDIR/staged to /mount/your/desired/partition/staged achieve what you want?

**Comment by pohly** [comment]
@icarus-sparry yes, I can mess around with the statedir and create a symlink like you said. I think an explicit command line option would be clear and avoid embedding such knowledge about swupd internals in the script calling swupd.

**Comment by matthewrsj** [comment]
Can you explain why you can't just specify --statedir instead of a separate --stagedir? As far as content size goes, most of it is put in staged/ anyways, so having the entire statedir on the other partition doesn't seem like a big hit to me.

**Comment by pohly** [comment]
The intended partitioning is "two small system partitions (active and inactive), large data partition". --statedir is also used for downloaded tar archives and manifest files. Putting those onto the target partition increases the space overhead that is needed for swupd on the target partition. Less overhead is better. So while it is technically feasible, it's not as good as it good be.

A better solution would use the data partition for all files which do not have to be on the system partition. The staging directory needs to be on the system partition, to ensure that directory rewriting remains efficient (no need to write files, just delete + rename).
